packages/yew-router/src/router.rs

```diff
@@ -1,232 +1,111 @@
+/// stating that no route can be matched.
+/// See the [crate level document][crate] for more information.
+pub struct Router<R: Routable + Clone + PartialEq + 'static> {
+    props: RouterProps<R>,
+    #[allow(dead_code)] // only exists to drop listener on component drop
+    route_listener: RouteListener,
+    route: Option<R>,
 }
 
-impl<STATE, SW> Component for Router<SW, STATE>
+impl<R> Component for Router<R>
 where
-    STATE: RouterState,
-    SW: Switch + Clone + 'static,
+    R: Routable + Clone + PartialEq + 'static,
 {
-    type Message = Msg<STATE>;
-    type Properties = Props<STATE, SW>;
+    type Message = Msg<R>;
+    type Properties = RouterProps<R>;
 
     fn create(props: Self::Properties, link: ComponentLink<Self>) -> Self {
-        let callback = link.callback(Msg::UpdateRoute);
-        let mut router_agent = RouteAgentBridge::new(callback);
-        router_agent.send(RouteRequest::GetCurrentRoute);
+        let route_listener = attach_route_listener(link.callback(Msg::UpdateRoute));
 
-        Router {
-            switch: Default::default(), /* This must be updated by immediately requesting a route
-                                         * update from the service bridge. */
+        Self {
             props,
-            router_agent,
+            route_listener,
+            route: current_route(),
         }
     }
 
-    fn update(&mut self, msg: Self::Message) -> ShouldRender {
+    fn update(&mut self, msg: Self::Message) -> bool {
         match msg {
             Msg::UpdateRoute(route) => {
-                let mut switch = SW::switch(route.clone());
-
-                if switch.is_none() {
-                    if let Some(redirect) = &self.props.redirect {
-                        let redirected: SW = (&redirect.0)(route);
-
-                        log::trace!(
-                            "Route failed to match, but redirecting route to a known switch."
-                        );
-                        // Replace the route in the browser with the redirected.
-                        self.router_agent
-                            .send(RouteRequest::ReplaceRouteNoBroadcast(
-                                redirected.clone().into(),
-                            ));
-                        switch = Some(redirected)
-                    }
-                }
-
-                self.switch = switch;
+                self.route = route;
                 true
             }
         }
     }
 
-    fn change(&mut self, props: Self::Properties) -> ShouldRender {
-        self.props = props;
-        true
+    fn change(&mut self, mut props: Self::Properties) -> bool {
+        std::mem::swap(&mut self.props, &mut props);
+        props != self.props
     }
 
-    fn view(&self) -> VNode {
-        match self.switch.clone() {
-            Some(switch) => (&self.props.render.0)(switch),
+    fn view(&self) -> Html {
+        match self.route.clone() {
```

Comment:
If we pass a reference to `self.route` into the `RenderFn`, we could get rid of the `Clone` trait bound entirely.
We can also get rid of `PartialEq` so we're only left with `Routable`.

At this point
```rust
#[derive(Routable)]
enum Test {
  // ...
}
```
is really all you need.
Wouldn't that be great?

Reply:
Yeah, that's great. Done that.